ocaml/xenopsd/pvs/pvs_setup.ml

```diff
@@ -0,0 +1,603 @@
+  let ofport intf =
+    let cmd = Cmd.(vsctl % "get" % "interface" % intf % "ofport") in
+    let* out = run cmd in
+    try Scanf.sscanf out {|%d|} ok
+    with _e -> error "%s: failed to scan port from '%s'" __FUNCTION__ out
+
+  let add_flow ?(debug = false) bridge flow =
+    let flow' = String.concat "," flow in
+    let cmd =
+      match debug with
+      | true ->
+          Cmd.(echo % bridge % flow')
+      | false ->
+          Cmd.(ofctl % "--strict" % "add-flow" % bridge % flow')
+    in
+    D.info "%s: %s" __FUNCTION__ (Cmd.to_string cmd) ;
+    OS.Cmd.(run_out cmd |> to_stdout)
+
+  let del_flows bridge extra =
+    let args = "del-flows" :: bridge :: extra |> List.map Cmd.v in
+    let cmd = Cmd.(List.fold_left add_args ofctl args) in
+    D.info "%s: %s" __FUNCTION__ (Cmd.to_string cmd) ;
+    OS.Cmd.(run_out cmd |> to_stdout)
+end
+
+(** 16bit msb and lsb of the index *)
+let interface_id vif =
+  let path = "/sys/class/net" // vif // "ifindex" in
+  let ( >> ) = Int64.shift_right in
+  let ( & ) = Int64.logand in
+  try
+    scanning path @@ fun io ->
+    Scanf.bscanf io {|%Ld|} @@ fun id ->
+    (* this is what the original code does but it looks suspicious: it
+       takes the lower 16 bit and upper 32 bit of id *)
+    ok Int64.(id, id >> 32 |> to_int, (id & 0xffffL) |> to_int)
+  with _ -> error "failed to scan %s" path
+
+module Const = struct
+  let rule_prio = 1000
+
+  let client2server = 1
+
+  let proxy2client = 2
+
+  let server2client = 3
+
+  let port_table = 101
+
+  let action_table = 102
+
+  let dir = "reg0"
+
+  let client_msb = "reg1"
+
+  let client_lsb = "reg2"
+
+  let nxm_dir = "NXM_NX_REG0"
+
+  let nxm_client_msb = "NXM_NX_REG1"
+
+  let nxm_client_lsb = "NXM_NX_REG2"
+end
+
+(** implement the add command *)
+let add debug dev vif priv hotplug =
+  D.debug "%s: %s %s %s" __FUNCTION__ vif priv hotplug ;
+  let p = Printf.sprintf in
+  let* vif_uuid = read (priv // "vif-uuid") in
+  let* () = pvs_is_running vif_uuid in
+  let* pvs_proxy_if = read (priv // "pvs-interface") in
+  let* mac = read (priv // "mac") in
+  let* pvs_server_count = read (priv // "pvs-server-num") >>= to_int in
+  let* bridge' = read (priv // "bridge") in
+  let* pvs_bridge = OVS.pvs_bridge bridge' in
+  D.debug "%s: pvs_bridge = %s" __FUNCTION__ pvs_bridge ;
+  let* proxy_mac = OVS.mac_addr pvs_proxy_if in
+  let* proxy_port = OVS.ofport pvs_proxy_if in
+  let* vm_port = OVS.ofport vif in
+  let* id, id_msb, id_lsb = interface_id vif in
+
+  let add_pvs_server bridge n =
+    D.debug "%s: bridge = %s" __FUNCTION__ bridge ;
+    let* ip_addrs = pvs_server_addrs ~n priv in
+    let* port_lo, port_hi = pvs_server_ports ~n priv in
+    let* () =
+      (* flows per server address *)
+      ip_addrs
+      |> iter @@ fun ip ->
+         let* nw_src = OVS.nw_src ip in
+         let* nw_dst = OVS.nw_dst ip in
+         let* udp = OVS.udp ip in
+         let* ip = OVS.ip ip in
+         let* () =
+           OVS.add_flow ~debug bridge
```

Comment:
The old script has useful code comments explaining each block of add-flow, which would be good to include.